### ISSUE TITLE
Rerun triggers not causing dependent steps to be rerun

### ISSUE THREAD

**Issue by powelleric** [snakemake/snakemake#3095]
**Snakemake version**
8.11.6

**Describe the bug**
In a 2 step workflow, if I run the workflow to completion, then realize I must re-run step 1, I expect that if I run step 1 with --force that step 2 will be run on the next invocation of the whole workflow.

**Minimal example**
Given this snakefile called `bad.snakefile`:

```
rule runall:
    input: "bye.txt"
    output: "hello.txt"
    shell: "echo hello > {output}"


rule realoutput:
    output: "bye.txt"
    shell: "echo bye > {output}"
```

- First run `snakemake --snakefile bad.snakefile --cores 1`.  It works successfully and runs both steps.  
- Next, rerun step 1 using `snakemake --snakefile bad.snakefile --cores 1 --force realoutput`.  This successfully runs the first step.  
- The bug happens when I then try to run the full workflow with the first command: `snakemake --snakefile bad.snakefile --cores 1`.  This should cause Step 2 to be re-run.  However, it says everything is complete.  But it is clear that the mtime for step 1 output is greater than for step 2 output.  This is not fixed by adding `--rerun-triggers=mtime`.

It works correctly if I do `snakemake --snakefile bad.snakefile --cores 1 --forcerun realoutput`.  However, this forces me to run all the steps at once.  I wish to re-run step 1 and then have a chance to check the outcome.  Then, I want to proceed with the re-run of the dependent steps.

**Comment by tbooth** [comment]
I believe this is the same bug as per #2626 and #2011, and in brief summary:

* The behaviour seen by @powelleric is correct/expected but not well documented (see for example https://bioinformatics.stackexchange.com/questions/19923/modification-time-does-not-trigger-a-snakemake-rule#answer-19927)
* The `--rerun-triggers=mtime` option is deffo broken.

The Snakemake code for deciding what to rerun is pretty knarly and easy to break (see for example the attempts to fix #2015) so I think in terms of development the first task is to make a big regression test suite of corner cases. I have it vaguely in the back of my mind that this work could be spun into some sort of masters project or GSoC project, but I've not put any flesh on the bones of that idea yet.

**Comment by powelleric** [comment]
@tbooth Thank you so much for your prompt response.  I tried to search for others complaining about this but did not find any of the ones you linked to.  I really hope that you can get this issue resolved.  Snakemake has been a wonderful tool for me over the years but this bug really hurts.

**Comment by tbooth** [comment]
Yes it's annoying. Depending on your application you may find that always using `--drop-metadata` is a viable workaround. This does increase the risk of leaving partial output files if your workflow crashes part-way, but liberal use of `shadow: "minimal"` (ie. use it for every rule unless there is a particular reason not to) will address that (and is arguably good practice anyway).

**Comment by powelleric** [comment]
Wonderful.  Thank you for those suggestions.  That is just the type of guidance I was needing.

**Comment by powelleric** [comment]
I found that using --max-checksum-file-size 1b essentially fixes my problems.  I just print out a little bit of logging information and it is enough.  It is enough to make the original `bad.snakefile` example work properly.
